apps/voting-disputable/contracts/DisputableVoting.sol

```diff
@@ -0,0 +1,745 @@
+        require(_canExecute(votes[_voteId]), ERROR_CAN_NOT_EXECUTE);
+        _unsafeExecuteVote(_voteId);
+    }
+
+    /**
+    * @dev Internal unsafe version of _executeVote that assumes you have already checked if the vote can be executed and exists
+    */
+    function _unsafeExecuteVote(uint256 _voteId) internal {
+        Vote storage vote_ = votes[_voteId];
+
+        vote_.executed = true;
+        vote_.disputableStatus = DisputableStatus.Closed;
+        _closeAgreementAction(vote_.actionId);
+
+        bytes memory input = new bytes(0); // TODO: Consider input for voting scripts
+        address[] memory blacklist = new address[](1);
+        blacklist[0] = address(_getAgreement());
+        runScript(vote_.executionScript, input, blacklist);
+        emit ExecuteVote(_voteId);
+    }
+
+    /**
+    * @dev Challenge a vote
+    * @param _voteId Identification number of the vote to be challenged
+    */
+    function _onDisputableActionChallenged(uint256 _voteId, uint256 /* _challengeId */, address /* _challenger */) internal {
+        Vote storage vote_ = votes[_voteId];
+        require(_canPause(vote_), ERROR_CANNOT_PAUSE_VOTE);
+
+        vote_.disputableStatus = DisputableStatus.Paused;
+        vote_.pausedAt = getTimestamp64();
+        emit PauseVote(_voteId);
+    }
+
+    /**
+    * @dev Allow a vote
+    * @param _voteId Identification number of the vote to be allowed
+    */
+    function _onDisputableActionAllowed(uint256 _voteId) internal {
+        Vote storage vote_ = votes[_voteId];
+        require(_isPaused(vote_), ERROR_VOTE_NOT_PAUSED);
+
+        vote_.disputableStatus = DisputableStatus.Active;
+        vote_.pauseDuration = getTimestamp64().sub(vote_.pausedAt);
+        emit ResumeVote(_voteId);
+    }
+
+    /**
+    * @dev Reject a vote
+    * @param _voteId Identification number of the vote to be rejected
+    */
+    function _onDisputableActionRejected(uint256 _voteId) internal {
+        Vote storage vote_ = votes[_voteId];
+        require(_isPaused(vote_), ERROR_VOTE_NOT_PAUSED);
+
+        vote_.disputableStatus = DisputableStatus.Cancelled;
+        vote_.pauseDuration = getTimestamp64().sub(vote_.pausedAt);
+        emit CancelVote(_voteId);
+    }
+
+    /**
+    * @dev Void an entry
+    * @param _voteId Identification number of the entry to be voided
+    */
+    function _onDisputableActionVoided(uint256 _voteId) internal {
+        _onDisputableActionAllowed(_voteId);
+    }
+
+    /**
+    * @dev Internal function to check if a vote can be executed. It assumes the queried vote exists.
+    * @return True if the given vote can be executed, false otherwise
+    */
+    function _canExecute(Vote storage vote_) internal view returns (bool) {
+        // If vote is already executed, it cannot be executed again
+        if (vote_.executed) {
+            return false;
+        }
+
+        // If the vote is paused or cancelled, it cannot be executed
+        if (_isPausedOrCancelled(vote_)) {
```

Comment:
Checking the paused state seems kind of redundant; the only time this check would be important with the other one is when the vote is cancelled, right?

Reply:
Actually `_isVoteOpen` is already handling this case